Add "Recent" file listing

icewind1991 opened the pull request:
Show a list of recently modified files and folders

![recent](https://i.imgur.com/pqnCkw0.png)
### On showing folders

How to show folders in the list is the most complex part of this PR, since the mtime (both on disk and in cache) of a folder gets changed when a file in the folder is modified simply just showing folders in the same way as files would result in the parent folder being shown together with the most recently changed file in that folder.

Ideally we would show folders based on their creation time but unfortunately we don't save that information, instead this PR uses the modified date of the least recently changed file in the folder which isn't perfect (especially for folders containing a single file) but it works decently.

I could also be better to hide most folders altogether in most cases
### On hiding folder contents

Another thing I was thinking about to make the recent list more useful is hiding the contents of newly uploaded folders, currently if you upload a large folder the contents of it will take over the entire recents list.
Instead it would be nice if the recents list would only show the newly uploaded folder, with the newly uploaded contents being implicitly grouped in the folder.
### Further work (maybe for nc+1)
- [x] unit tests
- [ ] make the list as useful as possible
  - [ ] maybe hide folders where they don't add value
  - [ ] hide content of folders where they don't add value
- [x] Icon for the files sidebar
- [ ] Maybe allow specifying the range mtimes to be searched for (show me files modified 2 days ago)

cc @jancborchardt @LukasReschke

mar1u5 commented:
Wow... Amazing 👊 
How is Google Drive, OneDrive or Dropbox handling folder contents?

Bugsbane commented:
Love, love, love this! I can create an icon if needed. My first thought though, is that a "reload" type icon would work just fine here though. I mean like an arrow going around in a circle.

MorrisJobke commented:
> What think the rest of @nextcloud/designers? My preference is something Nautilus-like (despite being a KDE fanboy) such as this.

cc @nextcloud/designers

LukasReschke commented on apps/files/lib/Controller/ApiController.php:
```diff
@@ -171,6 +180,36 @@ public function getFilesByTag($tagName) {
+	 *
+	 * @return DataResponse
+	 */
+	public function getRecentFiles() {
```
Unit test for this? 😉

icewind1991 replied:
In the todo :smile:

williambargent commented:
:+1: Looks good, Just need the Icon. @Bugsbane I like the idea of some sort of clock face. Would be awesome if we could get it to actually show the time of the last file which was modified (Would probably cause to many issues).
![time-512](https://cloud.githubusercontent.com/assets/11518756/16555197/51d654b6-41cb-11e6-8244-68ebe52f6a9f.png)

icewind1991 commented:
> How is Google Drive, OneDrive or Dropbox handling folder contents?

It gets all it's data from the cache, as long as changes are scanned they should show up

jancborchardt commented:
Ok, so how does this exactly compare to »Sort by modified date«?

I think the general thing is cool and all, I am just worried we're adding more and more stupid sections to the navigation without properly connecting everything. That is, I think a much better solution would be to:
- use »Sort by modified date« as default in the all files list
- of course remember if that sorting is changed, on a per-user basis
- (just to bring an example of another nav section we can integrate better) we should always sort the favorite files of that folder up top

icewind1991 commented:
> Ok, so how does this exactly compare to »Sort by modified date«?

The recents "tab" shows recent files across all folders, not just the current folder

Bugsbane commented:
This is exactly the same as how file browsers let you both sorry by date, but also see all of the files you've modified recently, regardless of location. I use both daily, for different purposes.

jancborchardt commented:
It was partly a rhetorical question. ;) What I meant to say is that my proposal above would already take care of a majority of cases for the »Recent« section, but in a less interface-filling and more integrated way.

Let's take small steps, and find how we can do more with less. Again sure, we can just add stuff, but that's the trap software falls into before people abandon it cause it's too complicated.

Bugsbane commented:
draft mockup of a clock style icon for recently modified files:
![recent-icon-1](https://cloud.githubusercontent.com/assets/401733/16640585/29a5db62-43c7-11e6-8360-437dc4e58edd.png)

jancborchardt commented:
@Bugsbane very nice! :) I'd only increase the size by 1px for each side because compared to the square icons around it looks a tad small. That's because it has the same width as the folder or share icons, but being a circle it needs a bit more size to appear visually sized correctly.

Bugsbane commented:
Ok, this is as large as it can go, while still fitting into the 16x16 png image size:
![recent-icon-1](https://cloud.githubusercontent.com/assets/401733/16650224/1e4c6418-440b-11e6-8643-38fab5db80eb.png)

If it's ok, just let me know and I'll svgo/trimage it and commit.

icewind1991 commented:
> `png`

master will no longer have fallback png's so no need to add a png version

> what path / filename should I commit the svg/png to?

You'll need to add the `nav-icon-recent` class (see apps/files/css/files.css line ~90) which points to the newly added icon

Bugsbane commented:
Ah, ok. It's being pulled in by css. Thanks. I just did a quick search on the commit changes page for .png and .svg and didn't see anything, and assumed you hadn't added the icon yet as it didn't exist. I'll add the class.

jancborchardt commented:
@Bugsbane looks great! Please make sure the circle is pixel-perfect, that is it should be either 1px or most probably 2px width. That will ensure it will look sharp and not blur.

Bugsbane commented:
Ok, I've submitted PR https://github.com/nextcloud/server/pull/340 and made sure that as much as is possible, that everything uses whole pixel stroke widths and aligns with the pixel grid. One thing puzzles me though. I could see the class pulling in the svg, but I didn't see anything referring to the PNG fallbacks, although I figure that these substitutions are probably done automatically by JS rather than being hard coded.

Bugsbane commented:
Oh, and I tried svgo on it, but no matter how I tried to modify the original SVG, svgo always butchered the image, completely removing the hands of the clock or worse. Accordingly I ran sourge over it instead, which worked fine.

icewind1991 committed:
fix test

icewind1991 commented:
@LukasReschke @rullzer @nickvergessen please review

jancborchardt commented:
@icewind1991 when just switching to the branch, you get a »No files in here« message. Which is weird cause I assumed the »Recent« value is just taken from the modified time. If this list is not populated by default, it will be quite confusing as soon as people update.

icewind1991 commented:
> I assumed the »Recent« value is just taken from the modified time.

It is, can you check if there is anything in the log or check the result of `apps/files/api/v1/recent` call

schiessle commented:
@icewind1991 Great stuff!

Question: What about the open checkboxes at the top. Are all of them addressed? 

From a first gut feeling I would keep folders out of the list. The file which changed in the folder will be shown anyway. So I don't see the reason to show (all?) parent folders of all recently changes files as well in this view.

jancborchardt commented:
@icewind1991 the view only seems to get populated once I mess with new files.
How do I check the API call? Just opening the URL gives me ` Access forbidden. CSRF check failed`

And another thing: The »Recent« view also has headers to sort by size and name, and least recently updated. That doesn’t really make sense. At the very least we should reset that when switching back from the view, or ideally completely disable sorting in that view.

icewind1991 commented:
> How do I check the API call?

Open your browser's dev tools, go to the nextwork tab, click on the "Recent" button and see what calls are being made

> And another thing: The »Recent« view also has headers to sort by size and name, and least recently updated. That doesn’t really make sense. At the very least we should reset that when switching back from the view, or ideally completely disable sorting in that view.

True, they are "left overs" from the way the view is being made, will see if I can get rid of it

schiessle commented:
- [ ] The texteditor no longer loads on this branch, even after rebasing it on master (4a963c21fff251a5ea105e6ec50e6d87dca1654e). Directly on master everything works fine

icewind1991 committed:
use limit instead of since when listing recent files

icewind1991 committed:
block user sorting in recent files

icewind1991 committed:
Fix storage id with storage jail

icewind1991 committed:
add recent files to node api

icewind1991 committed:
add recent files api endpoint

icewind1991 committed:
Add recent file listing

icewind1991 commented:
- Fixed texteditor
- Disallow sorting
- Fixed no files showing up in some cases

This is good for re-review ( @jancborchardt @schiessle @MorrisJobke ) any open tasks in the original post are things we can look into for a further release

MorrisJobke commented:
I tested it and it works 👍

MorrisJobke commented:
```
There was 1 failure:

1) OCA\Files\Tests\Controller\ViewControllerTest::testIndexWithRegularBrowser
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
                         'appname' => 'files'
-                        'script' => 'list.php'
+                        'script' => 'recentlist.php'
                         'order' => 2
                         'name' => 'Recent'
                         'active' => false
                         'icon' => ''
                     )
                     2 => Array (...)
                     3 => Array (...)
                     4 => Array (...)
                     5 => Array (...)
                     6 => Array (...)
                     7 => Array (...)
                 )
             )
             'l10n' => OC\L10N\L10N Object (...)
             'theme' => OC_Defaults Object (...)
         )
         'appContents' => Array (...)
     )
     'renderAs' => 'user'
     'appName' => 'files'
     'headers' => Array (...)
     'cookies' => Array ()
     'status' => 200
     'lastModified' => null
     'ETag' => null
     'contentSecurityPolicy' => OCP\AppFramework\Http\ContentSecurityPolicy Object (...)
 )

/drone/src/github.com/nextcloud/server/apps/files/tests/Controller/ViewControllerTest.php:318

```

icewind1991 committed:
fix test

williambargent commented:
:+1: tested and works, it is missing the favourites column, I think this should still be included in the view. Without it looks a bit odd transitioning from Files to Recent as everything shifts over to the left.

MorrisJobke commented:
> Without it looks a bit odd transitioning from Files to Recent as everything shifts over to the left.

The other views also don't have the favorite column. I would leave it like it is now.

williambargent commented:
For consistency, it might be worth reviewing and adding favourites to all the views.

mar1u5 commented:
Is this something we want to have for Nextcloud 10? @karlitschek

MorrisJobke commented:
> Is this something we want to have for Nextcloud 10? @karlitschek

I would say no, because this is clearly a feature and we only back port bug fixes.

karlitschek commented:
Nice. I agree that we shouldn't backport this for now

jancborchardt commented:
@williambargent can you open a new issue about adding the favorites column to all views? It should indeed be there for everything, including search results.
